src/main/kotlin/org/jitsi/jicofo/jigasi/JigasiDetector.kt

```diff
@@ -0,0 +1,117 @@
+import org.jitsi.xmpp.extensions.colibri.ColibriStatsExtension
+import org.json.simple.JSONObject
+import org.jxmpp.jid.EntityBareJid
+import org.jxmpp.jid.Jid
+
+/**
+ * [JigasiDetector] manages the pool of Jigasi instances by joining a "brewery" room that Jigasis connect to
+ * and publish their status in MUC presence.
+ * @author Damian Minkov
+ */
+class JigasiDetector(
+    xmppProvider: XmppProvider,
+    breweryJid: EntityBareJid
+) : BaseBrewery<ColibriStatsExtension>(
+    xmppProvider,
+    breweryJid,
+    ColibriStatsExtension.ELEMENT_NAME,
+    ColibriStatsExtension.NAMESPACE,
+    createLogger().apply { addContext("type", "jigasi") }
+) {
+    override fun onInstanceStatusChanged(jid: Jid, status: ColibriStatsExtension) {}
+    override fun notifyInstanceOffline(jid: Jid) {}
+
+    /**
+     * Selects a jigasi instance which supports transcription.
+     * @param exclude a list of excluded instances.
+     * @param preferredRegions a list of preferred regions.
+     * @return the JID of the selected instance, or  `null` if there are no available jigasis that satisfy the
+     * constraints (are not excluded and support transcription).
+     */
+    fun selectTranscriber(exclude: List<Jid>, preferredRegions: Collection<String>): Jid? =
+        selectJigasi(instances, exclude, preferredRegions, config.localRegion, transcriber = true)
+
+    /**
+     * Selects a jigasi instance which supports SIP.
+     * @param exclude a list of excluded instances.
+     * @param preferredRegions a list of preferred regions.
+     * @return the JID of the selected instance, or  `null` if there are no available jigasis that satisfy the
+     * constraints (are not excluded and support SIP).
+     */
+    fun selectJigasi(exclude: List<Jid>, preferredRegions: Collection<String>): Jid? =
+        selectJigasi(instances, exclude, preferredRegions, config.localRegion, transcriber = false)
+
+    val stats: JSONObject
+        get() = JSONObject().apply {
+            this["sip_count"] = getInstanceCount { it.supportsSip() }
+            this["sip_in_graceful_shutdown_count"] = getInstanceCount { it.supportsSip() && it.isInGracefulShutdown() }
+            this["transcriber_count"] = getInstanceCount { it.supportsTranscription() }
+        }
+
+    /**
+     * The companion object is necessary for the implicit call to this.createLogger() in the super constructor!
+     */
+    companion object {
+        @JvmStatic
+        fun selectJigasi(
+            instances: List<BaseBrewery<ColibriStatsExtension>.BrewInstance>,
+            exclude: List<Jid>,
+            preferredRegions: Collection<String>,
+            localRegion: String?,
+            transcriber: Boolean
+        ): Jid? {
+
+            val availableInstances = instances.filter {
+                !exclude.contains(it.jid) &&
+                    !it.isInGracefulShutdown() &&
+                    if (transcriber) it.supportsTranscription() else it.supportsSip()
+            }
```

Comment:
I don't feel strongly about this, but to me this reads a bit better with separate filter calls for each condition:
```
val availableInstances = instances
    .filter { !exclude.contains(it.jid) }
    .filter { !it.isInGracefulShutdown() }
    .filter { if (transcriber) it.supportsTranscription() else it.supportsSip() }
```

I also wondered if we should have the other functions pass a predicate instead of a boolean for the transcriber vs sip filter, so this function would become:

```
fun selectJigasi(
            instances: List<BaseBrewery<ColibriStatsExtension>.BrewInstance>,
            exclude: List<Jid>,
            preferredRegions: Collection<String>,
            localRegion: String?,
            additionalFilter: (BaseBrewery<ColibriStatsExtension>.BrewInstance) -> Boolean
        ): Jid? {

            val availableInstances = instances
                .filter { !exclude.contains(it.jid) }
                .filter { !it.isInGracefulShutdown() }
                .filter(additionalFilter)
...
```

And a calling function would be:
```
fun selectTranscriber(exclude: List<Jid>, preferredRegions: Collection<String>): Jid? =
        selectJigasi(instances, exclude, preferredRegions, config.localRegion) { it.supportsTranscription() }
```

But, again, just a thought.  I wouldn't call it a blocker.

Reply:
I did the first part. I'll leave the second part for another time, because JigasiSelectorTest uses this static method in a bunch of places and it's a bit awkward to expose the predicates and use them in java.